overlord/devicestate/devicestate_cloudinit_test.go

```diff
@@ -56,6 +59,169 @@ func (s *cloudInitBaseSuite) SetUpTest(c *C) {
+	// note that actually the devicemgr ensure only cares about having a grade
+	// for uc20, it doesn't use the grade for anything right now, the install
+	// handler code however does care about the grade, so here we just default
+	// to signed
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	s.makeModelAssertionInState(c, "canonical", "pc20-model", map[string]interface{}{
+		"display-name": "UC20 pc model",
+		"architecture": "amd64",
+		"base":         "core20",
+		"grade":        "signed",
+		"snaps": []interface{}{
+			map[string]interface{}{
+				"name":            "pc-kernel",
+				"id":              "pckernelidididididididididididid",
+				"type":            "kernel",
+				"default-channel": "20",
+			},
+			map[string]interface{}{
+				"name":            "pc",
+				"id":              "pcididididididididididididididid",
+				"type":            "gadget",
+				"default-channel": "20",
+			}},
+	})
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand:  "canonical",
+		Model:  "pc20-model",
+		Serial: "serial",
+	})
+}
+
+func (s *cloudInitSuite) SetUpTest(c *C) {
+	s.cloudInitBaseSuite.SetUpTest(c)
+
+	// make a uc16/uc18 style model assertion for the device
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	s.makeModelAssertionInState(c, "canonical", "pc-model", map[string]interface{}{
+		"architecture": "amd64",
+		"kernel":       "pc-kernel",
+		"gadget":       "pc",
+		"base":         "core18",
+	})
+	devicestatetest.SetDevice(s.state, &auth.DeviceState{
+		Brand:  "canonical",
+		Model:  "pc-model",
+		Serial: "serial",
+	})
+}
+
+func (s *cloudInitUC20Suite) TestCloudInitUC20CloudGadgetNoDisable(c *C) {
+	si := &snap.SideInfo{
+		RealName: "pc",
+		Revision: snap.R(1),
+		SnapID:   "pcididididididididididididididid",
+	}
+
+	// create a gadget snap in snapstate
+	st := s.o.State()
+	st.Lock()
+	snapstate.Set(st, "pc", &snapstate.SnapState{
+		SnapType: "gadget",
+		Current:  snap.R(1),
+		Active:   true,
+		Sequence: []*snap.SideInfo{si},
+	})
+	st.Unlock()
+
+	// create a cloud.conf file in the gadget snap's mount dir
+	gadgetDir := filepath.Join(dirs.SnapMountDir, "pc", "1")
+	c.Assert(os.MkdirAll(gadgetDir, 0755), IsNil)
+	c.Assert(ioutil.WriteFile(filepath.Join(gadgetDir, "cloud.conf"), nil, 0644), IsNil)
+
+	// pretend that cloud-init finished running
+	statusCalls := 0
+	r := devicestate.MockCloudInitStatus(func() (sysconfig.CloudInitState, error) {
+		statusCalls++
+		return sysconfig.CloudInitDone, nil
+	})
+	defer r()
+
+	restrictCalls := 0
+	r = devicestate.MockRestrictCloudInit(func(state sysconfig.CloudInitState, opts *sysconfig.CloudInitRestrictOptions) (sysconfig.CloudInitRestrictionResult, error) {
+		restrictCalls++
+		c.Assert(state, Equals, sysconfig.CloudInitDone)
+		c.Assert(opts, Not(IsNil))
+		// a gadget cloud.conf is not NoCloud
+		c.Assert(opts.DisableNoCloud, Equals, false)
+		return sysconfig.CloudInitRestrictionResult{
+			Action:     "disabled",
```

Comment:
the outcome here should be restrict, no?

Reply:
yes it should be restrict, and the reason this doesn't fail any unit tests is because I'm not checking the logged messages for these uc20 tests, I will add that here as well